### PR TITLE
[SPARK-46769][SQL] Fix type inferring for timestamps without time zone in JSON/CSV

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to revert inferring of the `TIMESTAMP_NTZ` type in the CSV/JSON datasources before the PR https://github.com/apache/spark/pull/43243.

### Why are the changes needed?
After the PR https://github.com/apache/spark/pull/43243, the TIMESTAMP_NTZ type inference in CSV/JSON datasource got 2 new guards which means TIMESTAMP_NTZ should be inferred either if:
1. the SQL config `spark.sql.legacy.timeParserPolicy` is set to `LEGACY` or
2. `spark.sql.timestampType` is set to `TIMESTAMP_NTZ`.

otherwise CSV/JSON should try to infer `TIMESTAMP_LTZ`.

Both guards are unnecessary because:
1. when `spark.sql.legacy.timeParserPolicy` is `LEGACY` that only means Spark should use a legacy Java 7- parser: `FastDateFormat` or `SimpleDateFormat`. Both parsers are applicable for parsing `TIMESTAMP_NTZ`.
2. when `spark.sql.timestampType` is set to `TIMESTAMP_LTZ`, it doesn't mean that we should skip inferring of `TIMESTAMP_NTZ` types in CSV/JSON, and try to parse the timestamp string value w/o time zone like `2024-01-19T09:10:11.123` using a LTZ format **with timezone** like `yyyy-MM-dd'T'HH:mm:ss.SSSXXX`. _The last one cannot match any NTZ values for sure._

### Does this PR introduce _any_ user-facing change?
Yes, it could.

### How was this patch tested?
By running the modified test suites:
```
$ build/sbt "test:testOnly *JsonInferSchemaSuite"
```
and the affected test suites:
```
$ build/sbt "test:testOnly *CSVLegacyTimeParserSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by MaxGekk** [comment]
also cc @Hisoka-X @sadikovi

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala [hunk 1]
```diff
@@ -202,11 +202,8 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
     // We can only parse the value as TimestampNTZType if it does not have zone-offset or
     // time-zone component and can be parsed with the timestamp formatter.
     // Otherwise, it is likely to be a timestamp with timezone.
-    val timestampType = SQLConf.get.timestampType
-    if ((SQLConf.get.legacyTimeParserPolicy == LegacyBehaviorPolicy.LEGACY ||
-        timestampType == TimestampNTZType) &&
-        timestampNTZFormatter.parseWithoutTimeZoneOptional(field, false).isDefined) {
-      timestampType
+    if (timestampNTZFormatter.parseWithoutTimeZoneOptional(field, false).isDefined) {
+      SQLConf.get.timestampType
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:206]: Restored to the state of https://github.com/apache/spark/pull/40022

**Review by Hisoka-X — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchemaSuite.scala [hunk 2]
```diff
@@ -267,7 +267,9 @@ class CSVInferSchemaSuite extends SparkFunSuite with SQLHelper {
   test("SPARK-45433: inferring the schema when timestamps do not match specified timestampFormat" +
     " with only one row") {
     val options = new CSVOptions(
-      Map("timestampFormat" -> "yyyy-MM-dd'T'HH:mm:ss"),
+      Map(
+        "timestampFormat" -> "yyyy-MM-dd'T'HH:mm:ss",
+        "timestampNTZFormat" -> "yyyy-MM-dd HH:mm:ss"),
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchemaSuite.scala:272]: To infer the STRING type, the input must not match to **both** formats: `TIMESTAMP` and `TIMESTAMP_NTZ`. Just set `timestampFormat` is not enough.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:206]: Does this assume the LTZ parser can parse NTZ values? But it isn't true if LTZ and NTZ have different parsing patterns?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala:206]: I think it's literally wrong to infer a value as LTZ type by using the NTZ parser.

**Pull request closed without merging**